feat: BigQuery Storage Write API integration

prash-mi opened the pull request:
Added `BigQueryTemplate.writeJsonStream(...)` method with it's supporting classes and implementations for writing JSON File's inputStream to BigQuery Storage using the storage write API.  This uses exactly once delivery semantics which are supported by the write API, this will guarantee that the data doesn't get duplicated.

elefeint commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryJsonDataWriter.java:
```diff
@@ -0,0 +1,148 @@
+
+  private final Object lock = new Object();
+
+  private final Logger logger = Logger.getLogger(this.getClass().getName());
```
Unllike with client libraries, the Spring modules use SLF4J abstraction for logging (JUL can still be used as one of the implementations). So this would be `org.slf4j.Logger` static field initialized as `LoggerFactory.getLogger(BigQueryJsonDataWriter.class)`.

elefeint commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java:
```diff
@@ -57,6 +75,10 @@ public class BigQueryTemplate implements BigQueryOperations {
 
   private Duration jobPollInterval = Duration.ofSeconds(2);
 
+  private static final int JSON_STREAM_WRITER_BATCH_SIZE = 1000; // write records in batches of 1000
```
It would be good to make this configurable by the user -- could you add a property to `GcpBigQueryProperties`, and inject it with a new setter in this class?

prash-mi replied:
sure, I have added `spring.cloud.gcp.bigquery.jsonWriterBatchSize` property. Also we are checking if it's set too low (or not set at all), in that case it will be defaulted to `JSON_STREAM_WRITER_BATCH_SIZE`

elefeint commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java:
```diff
@@ -159,16 +181,152 @@ public ListenableFuture<Job> writeDataToTable(
+  @Override
+  public ListenableFuture<WriteApiResponse> writeJsonStream(
+      String tableName, InputStream jsonInputStream, Schema schema) {
+    createTable(tableName, schema); // create table if it's not already created
```
This sounds like a relatively lengthy operation -- are there timeout configurations we can expose to the users later?

prash-mi replied:
Definitely, I have added it as a feedback @ go/bq-write-api-sc-gcp-integration. We will look into it later.

elefeint commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java:
```diff
@@ -159,16 +181,152 @@ public ListenableFuture<Job> writeDataToTable(
+    SettableListenableFuture<WriteApiResponse> writeApiFutureResponse =
+        new SettableListenableFuture<>();
+    CompletableFuture<Void> completableFuture =
+        CompletableFuture.runAsync(
```
Non-actionable, just a note for future -- we may want to allow users to inject a custom threadpool here, like we do for Pub/Sub. Although with BigQuery writing, I doubt there is as high of parallelism as with Pub/Sub.

prash-mi replied:
Good idea, captured it as a feedback @  go/bq-write-api-sc-gcp-integration

elefeint commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java:
```diff
@@ -159,16 +181,152 @@ public ListenableFuture<Job> writeDataToTable(
+    // register success and failure callback
+    writeApiFutureResponse.addCallback(
+        (res) -> {
+          completableFuture.cancel(true); // interrupt the completableFuture
```
Is this "just in case"? The success callback ideally will happen after the future value is set. Completable future javadoc says that calling `cancel` is a no-op if already completed, but is there an actual scenario in which canceling an already-completed or alread-failed future is needed?

prash-mi replied:
I think calling cancel isn't needed in the `success` callback, as we would have already set the value and the `runAsync` task would have gotten completed. Is it okay if I remove the `success` callback and just keep the failure callback with the `cancel` call?

elefeint replied:
I guess I am not sure why the callback is needed at all. Invocation of success callback means that ` writeApiFutureResponse.set(apiResponse);` got called successfully, which will lead to `completableFuture` completing. Invocation of error callback means that one of the `writeApiFutureResponse.setException(e);` calls was reached, again leading to `completableFuture`  completing correctly.
What is the state of `completableFuture` if the logic in the runnable completes successfully?

prash-mi replied:
Thanks @elefeint . I did some standalone testing for this and it appears that we do not need a success callback because of the reason you mentioned. However we might need a failure callback as the developer might call ` writeApiFutureResponse.cancel(true);` asynchronously and then we may want to interrupt the async operation which is writing the data. Please let me know if it looks good to you, we can later execute `asyncTask` on the developer's thread pool.

elefeint commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java:
```diff
@@ -159,16 +181,152 @@ public ListenableFuture<Job> writeDataToTable(
+  private WriteApiResponse getWriteApiResponse(String tableName, InputStream jsonInputStream)
+      throws DescriptorValidationException, IOException, InterruptedException {
+    WriteApiResponse apiResponse = new WriteApiResponse();
+    BigQueryWriteClient client = BigQueryWriteClient.create();
```
This should be done as part of autoconfiguration, similar to [`BigQuery` object](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/e04f755f75c98619c11fde42106cff81e95f1540/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/bigquery/GcpBigQueryAutoConfiguration.java#L71) for two reasons -- 
1) to allow users to customize their own `BigQueryWriteClient` and provide it as part of user configuration.
2) to allow us to provide properties for customization that will automatically be used to configure the object in autoconfiguration.

prash-mi replied:
This makes sense, I have made a note of it and will update the code accordingly

elefeint replied:
One more benefit is that it will make the code more testable -- if a `BigQueryWriteClient` object can come from the outside of the template, it can be mocked.

One thing I was not sure about -- is `BigQueryWriteClient` object meant to be reused between different write attempts (possibly to different tables)?

prash-mi replied:
Sure @elefeint , I have modified the code

elefeint commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryJsonDataWriter.java:
```diff
@@ -0,0 +1,148 @@
+  @GuardedBy("lock")
+  private RuntimeException error = null;
+
+  void initialize(TableName parentTable, BigQueryWriteClient client)
```
Is this object generally meant to be re-used between different table? From the usage in `BigQueryTemplate` it does not seem so. Would it make sense to make this object immutable, meant for one-time use against a single table? Right now, some very strange things would happen if `initialize()` is called the second time while the first write is in progress.

prash-mi replied:
Thanks for pointing this out, This is supposed to be initialised just once for the given table. I have added a parameterised constructor @ `BigQueryJsonDataWriter` to initialise class and have removed the `initialize()` method. Also, I have marked `streamWriter` as `final`

elefeint reviewed (changes requested):
First general pass. @burkedavison will review in more detail from Java perspective.

burkedavison commented on spring-cloud-gcp-samples/spring-cloud-gcp-bigquery-sample/src/main/java/com/example/WebController.java:
```diff
@@ -49,6 +54,99 @@ public ModelAndView renderIndex(ModelMap map) {
+      WriteApiResponse apiResponse = writeApiFuture.get();
+      if (apiResponse.isSuccessful()) {
+        message = "Successfully loaded data to " + tableName;
+      } else if (apiResponse.getErrors() != null && apiResponse.getErrors().size() > 0) {
```
Following Effective Java recommendations, could the contract of `getErrors()` be modified to always return a collection -- even if that collection is empty?

prash-mi replied:
sure. it will always return a `List<StorageError>` , even if empty

burkedavison commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryJsonDataWriter.java:
```diff
@@ -0,0 +1,148 @@
+        if (this.parent.error == null) {
+          StorageException storageException = Exceptions.toStorageException(throwable);
+          this.parent.error =
+              (storageException != null) ? storageException : new RuntimeException(throwable);
```
I'm coming in to this a little late in the game, but has there been discussion around the error handling design for this class?

Wrapping this Throwable in a RuntimeException and only saving the first one seems like it might create some limitations for the user of this class to get the full context of the problem -- particularly if they're performing a `catch` on a particular exception that may now be hidden one level down within the wrapper.

Should the caller of the BigQueryJsonDataWriter provide an onFailure callback? Or perhaps the BigQueryJsonDataWriter.append method should take the AppendCompleteCallback as a parameter? (In both cases: the goal being to delegate the error handling decisions one level higher.)

I'm not sure. Just for consideration.

prash-mi replied:
Makes sense and thanks for pointing this. I think we will need to improve the error handling design here. Say a user is writing data in bulk, like billions of records and say append failed for a few thousand. Then we should be having a mechanism to pass this message back to the user without interrupting the whole operation (as we might be still good as far as analytics is concerned).This same applies while committing the records. Let me take it back to the drawing board and address this in a followup PR.  Let me have a word with @elefeint  as well on this.

prash-mi replied:
captured as an AI @ go/bq-write-api-sc-gcp-integration

prash-mi commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java:
```diff
@@ -159,16 +181,152 @@ public ListenableFuture<Job> writeDataToTable(
+
+    BigQueryJsonDataWriter writer = new BigQueryJsonDataWriter();
+    // Initialize a write stream for the specified table.
+    writer.initialize(parentTable, client);
```
Thanks for having a look @burkedavison , I have refactored the class and have implemented `AutoClosable` as well.

prash-mi commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryJsonDataWriter.java:
```diff
@@ -0,0 +1,148 @@
+    private final BigQueryJsonDataWriter parent;
+
+    public AppendCompleteCallback(BigQueryJsonDataWriter parent) {
+      this.parent = parent;
```
sure. Refactored it accordingly

prash-mi commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryJsonDataWriter.java:
```diff
@@ -0,0 +1,148 @@
+        JsonStreamWriter.newBuilder(writeStream.getName(), writeStream.getTableSchema()).build();
+  }
+
+  public void append(JSONArray data, long offset)
```
Sure, added javadoc

prash-mi commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java:
```diff
@@ -159,16 +181,152 @@ public ListenableFuture<Job> writeDataToTable(
+
+    // set isSucccessful flag to true of there were no errors
+    if (apiResponse.getErrors().size() == 0) {
+      apiResponse.setSuccessful(true);
```
sure, I have captured this as an AI @ go/bq-write-api-sc-gcp-integration and I will address it in a followup PR if it's okay

elefeint commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryJsonDataWriter.java:
```diff
@@ -0,0 +1,156 @@
+    synchronized (this.lock) {
+      // If earlier appends have failed, we need to reset before continuing.
+      if (this.error != null) {
+        throw this.error;
```
Should the error get reset based on comment above? It seems to be unrecoverable based on usage in the template.

prash-mi replied:
We get to know about the failure asynchronously @ `onFailure` and we throw this exception when the developer tries to append @ `BigQueryTemplate`. I have captured a task in the design document to better handle the errors (on the lines of what we had discussed, like returning the unique set of exceptions and the count of exceptions). Is it okay if we plan to address this in a follow-up PR ?

elefeint commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryJsonDataWriter.java:
```diff
@@ -0,0 +1,156 @@
+   *
+   * @param client BigQueryWriteClient
+   */
+  public void finalizeWriteStream(BigQueryWriteClient client) {
```
Since the client got passed in the constructor, would it make sense to persist it there as a field, and have `finalizeWriteStream()` method be no-argument?

prash-mi replied:
sure. updated the code

elefeint commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryJsonDataWriter.java:
```diff
@@ -0,0 +1,156 @@
+    // Verify that no error occurred in the stream.
+    synchronized (this.lock) {
+      if (this.error != null) {
+        throw this.error;
```
When errors get thrown in different places, does `close()` method need to be called to clean up the writer?

elefeint commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryJsonDataWriter.java:
```diff
@@ -0,0 +1,156 @@
+
+    public void onSuccess(AppendRowsResponse response) {
+      logger.info(
+          String.format("\nAppend %d success", response.getAppendResult().getOffset().getValue()));
```
String.format() has a cost, but as it happens, slf4j `logger` has overloaded method variants that take a string pattern and parameters. This way, the string substitution will only be done if `INFO` level is needed.

prash-mi replied:
done

elefeint commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java:
```diff
@@ -159,16 +188,164 @@ public ListenableFuture<Job> writeDataToTable(
+      BufferedReader jsonReader = new BufferedReader(new InputStreamReader(jsonInputStream));
+      String jsonLine = null;
+      JSONArray jsonBatch = new JSONArray();
+      while ((jsonLine = jsonReader.readLine()) != null) { // read the input stream line by line
```
Will every line of input be required to be a valid JSON object? What about fixed-length formatted input, or minimized input in which all objects are on a single line?

prash-mi replied:
Do we want to restrict to newline delimited JSON format and then we can explore fixed-length formatted input, or minimised input as per the user's feedback. BQ Write API supports newLine Separated JSON out of the box (https://cloud.google.com/bigquery/docs/loading-data-cloud-storage-json) and we may have to do additional processing at the client library layer for any additional formats we add, which we can plan to add as per the user's demand. WDYT?

elefeint reviewed (changes requested):
The bad news about why there are no unit tests in the BigQuery integration is that [it was too difficult to unit test](https://github.com/spring-attic/spring-cloud-gcp/pull/1916#discussion_r327685714).

However, a lot of the code you are adding lends itself better to being unit tested -- for example, with `jsonWriterBatchSize`, setting it and making sure that the correct batch size is passed to the client object. Do what you can for the unit testing of `BigQueryTemplate` and `GcpBigQueryAutoConfiguration`, even if it won't reach the normal 80% threshold.

prash-mi commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java:
```diff
@@ -57,6 +75,17 @@ public class BigQueryTemplate implements BigQueryOperations {
+
+  // @Value("${spring.cloud.gcp.bigquery.jsonWriterBatchSize}") TODO(prasmish) debug exception while
+  // autowiring
+  private int jsonWriterBatchSize;
```
sure. I have now autowired accordingly

prash-mi commented:
> The bad news about why there are no unit tests in the BigQuery integration is that [it was too difficult to unit test](https://github.com/spring-attic/spring-cloud-gcp/pull/1916#discussion_r327685714).
> 
> However, a lot of the code you are adding lends itself better to being unit tested -- for example, with `jsonWriterBatchSize`, setting it and making sure that the correct batch size is passed to the client object. Do what you can for the unit testing of `BigQueryTemplate` and `GcpBigQueryAutoConfiguration`, even if it won't reach the normal 80% threshold.

Sure. Let me update you after adding unit tests.

prash-mi commented:
> > The bad news about why there are no unit tests in the BigQuery integration is that [it was too difficult to unit test](https://github.com/spring-attic/spring-cloud-gcp/pull/1916#discussion_r327685714).
> > However, a lot of the code you are adding lends itself better to being unit tested -- for example, with `jsonWriterBatchSize`, setting it and making sure that the correct batch size is passed to the client object. Do what you can for the unit testing of `BigQueryTemplate` and `GcpBigQueryAutoConfiguration`, even if it won't reach the normal 80% threshold.
> 
> Sure. Let me update you after adding unit tests.

@elefeint JFYI I have added UT for the new methods @ `BigQueryTemplate` and `GcpBigQueryAutoConfiguration`. Currently getting ~ 40% coverage

elefeint reviewed (changes requested):
Broadly the PR is looking great -- thank you for adding the tests! I've run through the sample, and that's very good, too.

I have some more minor comments inlined. 

Also, could you update [CHANGELOG](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/main/CHANGELOG.md) and take a look at Sonar code smells?

prash-mi commented:
> Broadly the PR is looking great -- thank you for adding the tests! I've run through the sample, and that's very good, too.
> 
> I have some more minor comments inlined.
> 
> Also, could you update [CHANGELOG](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/main/CHANGELOG.md) and take a look at Sonar code smells?

@elefeint . Sure, I see the last released version was `3.3.0`, I have added v `3.3.1` in the change log. Please let me know if it looks good. Also I have addressed the Sonar code smells

prash-mi commented on spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/bigquery/GcpBigQueryAutoConfiguration.java:
```diff
@@ -80,7 +89,24 @@ public BigQuery bigQuery() throws IOException {
+    BigQueryWriteSettings bigQueryWriteSettings =
+        BigQueryWriteSettings.newBuilder()
+            .setCredentialsProvider(
+                FixedCredentialsProvider.create(this.credentialsProvider.getCredentials()))
```
sure, Done

prash-mi commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java:
```diff
@@ -159,16 +222,194 @@ public ListenableFuture<Job> writeDataToTable(
+    // Initialize a write stream for the specified table.
+    BigQueryJsonDataWriter writer = getBigQueryJsonDataWriter(parentTable);
+
+    // new BigQueryJsonDataWriter(parentTable, bigQueryWriteClient);
```
sure, done

prash-mi commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/WriteApiResponse.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class WriteApiResponse {
```
But we will be required to access it outside the package, for example our `com.example` package uses `WriteApiResponse`.  Also, the developers might use the public `ListenableFuture<WriteApiResponse> writeJsonStream` method in `BigQueryTemplate` and hence they will also need to access `WriteApiResponse` in their own packages. Please let me know if I am missing anything here.

elefeint replied:
That makes sense; it was meant to be part of public API. Thanks!

prash-mi commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java:
```diff
@@ -51,38 +70,82 @@ public class BigQueryTemplate implements BigQueryOperations {
    */
-  public BigQueryTemplate(BigQuery bigQuery, String datasetName) {
-    this(bigQuery, datasetName, new DefaultManagedTaskScheduler());
+  public BigQueryTemplate(
```
This makes sense @elefeint , I have marked the two constructors ` @Deprecated` and we can use  `BigQueryTemplate(BigQuery bigQuery, BigQueryWriteClient bigQueryWriteClient, Map<String, Object> bqInitSettings, TaskScheduler taskScheduler)` .

prash-mi commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java:
```diff
@@ -51,38 +70,82 @@ public class BigQueryTemplate implements BigQueryOperations {
+    this(
+        bigQuery,
+        bigQueryWriteClient,
+        (String) bqInitSettings.get("DATASET_NAME"),
```
done

elefeint reviewed (approved):
Thank you very much! I'm sorry it took so long to review.

I'll merge the PR in, and move the changelog from 3.3.1 to 3.4.0-SNAPSHOT (the next release will be 3.4.0).

It would also be good to add documentation in a separate PR -- unfortunately right now we have to duplicate documentation between `bigquery.adoc` and `bigquery.md` in preparation for doc publishing migration.

prash-mi commented:
@elefeint Thanks for merging the PR. I will add documentation in both the files (`bigquery.adoc` and `bigquery.md`)

meltsufin commented on spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java:
```diff
@@ -159,16 +226,187 @@ public ListenableFuture<Job> writeDataToTable(
+        new SettableListenableFuture<>();
+
+    Thread asyncTask =
+        new Thread(
```
It seems dangerous to create an unbounded number of threads without a thread pool here. It might be causing the OutOfMemoryError reported in #1599.

cc: @prash-mi, @burkedavison

burkedavison replied:
+1 : Stack trace in the issue points to [asyncTask.start()](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/v3.4.0/spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java#L310)

Good stackoverflow on the issue here: https://stackoverflow.com/a/16789621
